marlowe-spec-test/src/Marlowe/Spec/Core/Guarantees.hs

```diff
@@ -0,0 +1,332 @@
+    , computeTransactionIsQuiescentTest i
+    , playTraceIsQuiescentTest i
+    -- Timeout.thy
+    , timedOutTransaction_closes_contractTest i
+    -- CloseIsSafe.thy
+    , closeIsSafeTest i
+    ]
+
+-- TransactionBound.thy
+--
+-- lemma playTrace_only_accepts_maxTransactionsInitialState :
+--    "playTrace sl c l = TransactionOutput txOut ⟹
+--      length l ≤ maxTransactionsInitialState c"
+playTrace_only_accepts_maxTransactionsInitialStateTest :: InterpretJsonRequest -> TestTree
+playTrace_only_accepts_maxTransactionsInitialStateTest interpret = reproducibleProperty "lemma playTrace_only_accepts_maxTransactionsInitialState" do
+    context <- run $ generateT $ genContext interpret
+    (contract, Arith.Int_of_integer startTime, transactions) <- run $ generate $
+        frequency
+          [ (5, genContractStartTimeAndValidTransactions context)
+          , (20, genContractStartTimeAndValidTransactions context `suchThat` \(_,_,l) -> not (null l))
+          , (75, genContractStartTimeAndValidTransactions context `suchThat` \(_,_,l) -> length l > 1)
+          ]
+    let
+        req :: Request JSON.Value
+        req = PlayTrace startTime contract transactions
+    RequestResponse res <- run $ liftIO $ interpret req
+
+    case JSON.fromJSON res of
+      JSON.Success (TransactionOutput TransactionOutputRecord_ext {}) -> do
+        monitor
+          ( counterexample $
+              "Request: " ++ showAsJson req ++ "\n"
+                ++ "Expected maxTransactionsInitialState (" ++ show (maxTransactionsInitialState contract) ++ ")\n"
+                ++ "to be an upper bound for the number of transactions (" ++ show (length transactions) ++ ")")
+        stop $ cover 50.0 (length transactions > 1) "more than one transaction" $ toInteger (length transactions) <= integer_of_nat (maxTransactionsInitialState contract)
+        :: PropertyM ReproducibleTest Bool
+      JSON.Success (TransactionError err ) -> fail $ "Unexpected Transaction Error: " ++ show err
+      _ -> fail "JSON parsing failed!"
+
+genContractStartTimeAndValidTransactions :: Context -> Gen (Contract, Arith.Int, [Transaction_ext ()])
+genContractStartTimeAndValidTransactions context = do
+    c <- semiArbitrary context `suchThat` (/=Close)
+    s <- arbitraryPositiveInteger
+    t <- arbitraryValidTransactions (emptyState s) c
+    pure (c, s, t)
+
+isWhen :: Contract -> Bool
+isWhen When {} = True
+isWhen _ = False
+
+isClose :: Contract -> Bool
+isClose Close = True
+isClose _ = False
+
+-- SingleInputTransactions.thy
+--
+-- theorem traceToSingleInputIsEquivalent:
+--    "playTrace sn co tral = playTrace sn co (traceListToSingleInput tral)"
+traceToSingleInputIsEquivalentTest :: InterpretJsonRequest -> TestTree
+traceToSingleInputIsEquivalentTest interpret = reproducibleProperty "theorem traceToSingleInputIsEquivalent" do
+    (contract, Arith.Int_of_integer startTime, transactions) <- run $ do
+      generateT $ (do
+        context <- genContext interpret
```

Comment:
Only occasionally I observed timeouts of this test, which was due to the fact, that with a given `context` it was not possible to find suitable `transactions`. For now, I'm just discarding the `context` as well when the condition for the transactions does not hold. Therefore the test now takes longer. It might be worth investigating the "quality" of the `context` as well.